### ISSUE TITLE
A couple of hints

### ISSUE THREAD

**Issue by amivaleo** [Maestroschan/emoji-selector-for-gnome#20]
Hi! 

I would like to give you a couple of hints and maybe some ideas. I do that also because I'd like to understand more about extension-scripting.
Consider this screenshot. I will refer to it.
![schermata da 2017-11-01 11-38-08](https://user-images.githubusercontent.com/24316547/32272972-c3bed4f0-bf00-11e7-89ba-5319e0bf30fb.png)

- I updated the italian translation. I tried to use a more literal translation. Some entry was missing in the old po file (like the "Type here to search..." sentence in #1). I made a fork for that, but I'm not going to make a pull request right now. I'd like to play a bit with the other files first.

- I noticed that, if I do not install any noto/emojione/twitter/etc font, all the emojis in the default set have a differente width (#5). This makes every row (you called that "ln" in your extension.js file: `ln = new PopupMenu.PopupBaseMenuItem...`) of different width.
You could avoid that by adding some "width: # em" and "height: #em" to the stylesheet file. Something like:
```
.EmojisItemStyle {
    font-family: EmojiOne Color, Twitter Color Emoji, Noto Emoji;
    padding: 1px 3px;
    height: 1.8em;
    width: 1.8em;
}
```

- There is some padding/margin between each row of emojis. You could style that if you change this line:
`ln = new PopupMenu.PopupBaseMenuItem({reactive: true});`
to this:
`ln = new PopupMenu.PopupBaseMenuItem({style_class: 'EmojisList', reactive: true});`
and then add something like
```
.EmojisList {
    padding: 0px;
    margin: 0px;
}
```
in your stylesheet.css.

- I do not understand why the recent emojis are 4px bigger o.O If you remove that "+ 4" in this line
`let fontStyle = 'font-size: ' + ( Convenience.getSettings().get_int('emojisize') + 4 ) + 'px;'`
that should reduce/avoid the asymmetric padding (#3) in the submenus.

And... A few question for you:
- Can't you just remove the row #2 in every submenu? They do not make too much sense in my opinion! ^_^" but maybe it's just my opinion after all.

- Is it possible to give a fixed height (#4) to all submenus? It keeps changing for all of them.

Ok, that's enough for now. :o
Sorry, I am too much talkative.

**Comment by Maestroschan** [comment]
Concerning CSS you know more than me, i will let you try it in your fork and wait for the pull request.

> I do not understand why the recent emojis are 4px bigger

Because these emojis are a lot more used, so they need to be bigger (and i didn't want an other setting for this particular detail). I know, it's a little dishonest, the user set 28 but have 32 :smile:
I assumed that symetric padding would still be here if the size was the same, because of the scrollbar.
Now that research is functional, maybe categories' emojis don't need to be smaller anymore so i can stop this behavior...

> Can't you just remove the row #2 in every submenu?

I was not sure icons were clear enough (i'm not even sure they are all on the system of everyone...), and buttons don't have a different style when clicked, so i let this label visible for now.
When the clicked button will have a different style, i will make it invisible.

> Is it possible to give a fixed height (#4) to all submenus? It keeps changing for all of them.

No, it would mean wasted empty space for some categories, or too much scrolling for others (+ it depends on number of columns set and user's screen)

**Comment by amivaleo** [comment]
> i will let you try it in your fork and wait for the pull request.
I did something already, that's how I took that screenshot in my previous message. But I think I mixed different version of your repo, so... I'll fix that first, then I'll play with it and, finally, I'll bother you! :D

> I assumed that symetric padding would still be here if the size was the same
It's just that... Things could get really meesed up if the user wants to go retarded! :D
Look:
![senzanome](https://user-images.githubusercontent.com/24316547/32277094-a8f3b18e-bf11-11e7-9265-675d9ab3e2ce.png)

> I was not sure icons were clear enough (i'm not even sure they are all on the system of everyone...)
Uhm...
I switched to ubuntu 17.10 from ubuntu 16.04 yesterday. I had to pay attention on those icons there. :3
What I mean is that they are shown if the right font is installed by default. This naturally leads to the question "what are the default fonts installed by gnome on every distro?". I can't answer, but... I can play with several virtual machines. It will take time...

> and buttons don't have a different style when clicked
Aaah!
How dare you!? This is like a sword in my heart! D:
Now I've noticed that. I have to fix it. D:

> No, it would mean wasted empty space for some categories
Ok, but I just wanted to know if it is feasible and how. 😄
I'm not telling you that you should do that. I want to learn something... Please. :3

Anyway, I'm quite happy with extension. Thank you for all your effort. :)
As you can see, I like whining, but I much appreciate the fact that you can bear all this whining. :)

**Comment by Maestroschan** [comment]
oops i wrote symetric instead of asymetric...

> I switched to ubuntu 17.10 from ubuntu 16.04 yesterday. I had to pay attention on those icons there. :3
> What I mean is that they are shown if the right font is installed by default. This naturally leads to the question "what are the default fonts installed by gnome on every distro?". I can't answer, but... I can play with several virtual machines. It will take time...

No, icons are a matter of icon themes, i should have just look in /usr/share/icons/Adwaita but i didn't do it because i forgot. Now I only provide Caffeine icon because it was not in my current theme, and flag icon from the GNOME minesweeper game

> Ok, but I just wanted to know if it is feasible and how. ￼ :smile: 

The first solution which comes to my mind is adding some empty PopupBaseMenuItem. Their total number per category should be something like (the maximum number of emojis in a category)/(the number of column set by the user)

**Comment by amivaleo** [comment]
> No, icons are a matter of icon themes

Oh, you're right! I thought you used some character for the circle buttons. Now I see the lines in which you put the names of some symbolic icon.
Well... I'd say your choice is good. I only have some problem with the "emblem" icon. I didn't recognize it. Can't you just user the heart icon instead? It's also a very commonly used emoji, so it is good to show where you can find it.

> Now I only provide Caffeine icon because it was not in my current theme, and flag icon from the GNOME minesweeper game
"preferences-desktop-locale-symbolic" could be good as a flag icon. But for the drink&food icon... Nope, there's nothing in adwaita for that.

> The first solution which comes to my mind is adding some empty PopupBaseMenuItem

I'll find a way to learn. Drawing mockups is good but it's not enough anymore. :|

**Comment by Maestroschan** [comment]
I'm doing various things on the extension right now, i've copy-pasted what you wrote in your first post here but 1.8em is too much, i'll use 1.2em

**Comment by amivaleo** [comment]
Uhm... This is what happens if I put 1.2em. o.O
![senzanome](https://user-images.githubusercontent.com/24316547/33242262-281d0c98-d2d2-11e7-8d12-6aafa9259a70.png)

How do you refresh the extension on a wayland session? I used to restart the shell, but you can't here.

**Comment by Maestroschan** [comment]
I guess it depends on the font https://i.imgur.com/04iJ6i1.png the only emoji which doesn't fit on my system is 😴 (because it's larger than all other emojis for some reason). What's the minimum value working with your font ?

----

No you can't reload with Wayland, so i still use Xorg just for that. However, if it's just for the CSS, "Alt+F2 rt" should work even with Wayland.

**Comment by amivaleo** [comment]
I'm using noto emoji. I suppose that each emoji in that font has the same size, because I have no problem with any of them in particular.

1.3em is the minimum value at which I still can see the emoji.
Look at the screen: I set 1.2em in the ".EmojisItemStyle" css class, while I have 1.3em in the ".RecentItemStyle" class. As you can see, the recently used emojis are still visible thanks to that 1.3.
![schermata da 2017-11-26 20-14-44](https://user-images.githubusercontent.com/24316547/33243436-768f6d30-d2e6-11e7-8334-eea81ed5d278.png)

So... That's how it works? You can only restart the shell to see the changes you committing in your extension? o.o That's bad.

**Comment by Maestroschan** [comment]
I did some commits yesterday, everything is in the same CSS class now.
I will use 1.4 or 1.5 because nice fonts have coherent and correctly sized emojis but default fonts are problematic even with 1.3 on an Ubuntu 17.10 VM

**Comment by amivaleo** [comment]
I must apologize, because I use a full HD resolution on a 13 inches display. Things are very small here, that's why I exaggerate the values in the CSS, so that they look big enough for my finger! :)

**Comment by Maestroschan** [comment]
You don't have to apologize, it should look great on all devices and not only mine, it's cool to test the extension on several devices

**Comment by Maestroschan** [comment]
I'll use 1.5em but remove the padding i think

**Comment by amivaleo** [comment]
That could be ok. Although I like to have a bigger spacing in between. I suppose it's due to the fact that in my ubuntu 17.10 emojis are not coloured. I tried to follow some guide (like this: https://steemit.com/emoji/@bgv/color-emoji-on-ubuntu )  but they just stay grey.

Anyway, I was wondering: is there a way to make a while (or for) loop all over the available emojis in a font? If it were possible, you wouldn't need the emojisCharacters.js file.
I ask this because I noticed that even noto-emoji-font is not a complete emoji set:
![schermata da 2017-11-27 11-03-26](https://user-images.githubusercontent.com/24316547/33261343-ce288816-d362-11e7-9b6a-2d661ba75217.png)
